gtsam/linear/HessianFactor.cpp

```diff
@@ -72,7 +72,7 @@ HessianFactor::HessianFactor() :
 
 /* ************************************************************************* */
 HessianFactor::HessianFactor(Key j, const Matrix& G, const Vector& g, double f)
-    : GaussianFactor(KeyVector{j}), info_(Dims{G.cols(), 1}) {
+    : GaussianFactor(KeyVector{j}), info_(Dims{static_cast<Key>(G.cols()), 1}) {
```

Comment:
`Dims` is consumed by
```c++
    /// Construct from a container of the sizes of each block.
    template<typename CONTAINER>
    SymmetricBlockMatrix(const CONTAINER& dimensions, bool appendOneDimension = false) :
      blockStart_(0)
    {
      fillOffsets(dimensions.begin(), dimensions.end(), appendOneDimension);
      matrix_.resize(variableColOffsets_.back(), variableColOffsets_.back());
      assertInvariants();
    }
```
which is templated and will be inferred to ` SymmetricBlockMatrix(const vector<Key>&)`.

which is then
```c++
    template<typename ITERATOR>
    void fillOffsets(ITERATOR firstBlockDim, ITERATOR lastBlockDim, bool appendOneDimension)
    {
      variableColOffsets_.resize((lastBlockDim-firstBlockDim) + 1 + (appendOneDimension ? 1 : 0));
      variableColOffsets_[0] = 0;
      DenseIndex j=0;
      for(ITERATOR dim=firstBlockDim; dim!=lastBlockDim; ++dim) {
        variableColOffsets_[j+1] = variableColOffsets_[j] + *dim;
        ++ j;
      }
      if(appendOneDimension)
      {
        variableColOffsets_[j+1] = variableColOffsets_[j] + 1;
        ++ j;
      }
    }
```
where everything becomes `Eigen::DenseIndex` which should be the same as `Eigen::Index`.

Reply:
In the current develop branch, for HessianFactor and JacobianFactor, when you pass a gtsam::Key, their constructor will cast it to Eigen::Index. If someone is ```using gtsam::symbol_shorthand;``` or ```gtsam::Symbol```, these values might be larger than what Eigen::Index can hold.

```
static const size_t keyBits = sizeof(Key) * 8;
static const size_t chrBits = sizeof(unsigned char) * 8;
static const size_t indexBits = keyBits - chrBits;
static const Key chrMask = Key(UCHAR_MAX)  << indexBits; // For some reason, std::numeric_limits<unsigned char>::max() fails
static const Key indexMask = ~chrMask;
```

```
Symbol(unsigned char c, std::uint64_t j)
```


In the Macros.h of Eigen, you can find
```
#define EIGEN_DEFAULT_DENSE_INDEX_TYPE std::ptrdiff_t
typedef EIGEN_DEFAULT_DENSE_INDEX_TYPE Index;
```

For std::ptrdiff_t, it's not guaranteed to be 64 bit.
ref: https://en.cppreference.com/w/cpp/types/ptrdiff_t

Say for example 32bit ARM device or Emscripten. On these platform std::ptrdiff_t is smaller than from the definition in types.h
```
  typedef std::uint64_t Key;
```

It's not very straightforward to add these platforms for testing, as they are cross compiled and might not be able to run on the build system.

I haven't notcied any issue in ```SymmetricBlockMatrix.h``` but for safety it should be changed to gtsam::Key as well.

Reply:
I alos don't understand the change above. A dimension is not a key. And as Fan says, it will go to Eigen::Index.